stdlib/public/core/StringLegacy.swift

```diff
@@ -213,6 +213,65 @@ extension String {
 // rdar://problem/18878343
 #endif
 
+@_inlineable // FIXME(sil-serialize-all)
+@_versioned // FIXME(sil-serialize-all)
+internal func _ascii8(_ c: Unicode.Scalar) -> UInt8 {
+  _sanityCheck(c.value >= 0 && c.value <= 0x7F, "not ASCII")
+  return UInt8(c.value)
+}
+
+@_inlineable // FIXME(sil-serialize-all)
+@_versioned // FIXME(sil-serialize-all)
+internal func _digitASCII(
+  _ digit: UInt8, numeralsOnly: Bool, uppercase: Bool
+) -> UInt8 {
+  if numeralsOnly || digit < 10 {
+    return _ascii8("0") &+ digit
+  } else {
+    let base = (uppercase ? _ascii8("A") : _ascii8("a")) &- 10
+    return base &+ digit
+  }
+}
+
+@_inlineable // FIXME(sil-serialize-all)
+@_versioned // FIXME(sil-serialize-all)
+internal func _integerToString<T: FixedWidthInteger>(
+  _ value: T, radix: Int, uppercase: Bool
+) -> String {
+  if value == 0 {
+    return "0"
+  }
+
+  // Bit shifting / masking is much faster than division when `radix`
+  // is a power of two.
+  let radixIsPowerOfTwo = radix.nonzeroBitCount == 1
+  let radix = T.Magnitude(radix)
+  let quotientAndRemainder: (T.Magnitude) -> (T.Magnitude, T.Magnitude) =
+    radixIsPowerOfTwo
+      ? { ( $0 &>> radix.trailingZeroBitCount, $0 & (radix - 1) ) }
+      : { $0.quotientAndRemainder(dividingBy: radix) }
+
+  let isNegative = T.isSigned && value < 0
+  var value = value.magnitude
+
+  var result: [UInt8] = []
```

Comment:
Maybe worth calling `reseveCapacity` here? It's a really a nano-optimization, but may make a difference even if the capacity is, say, 4.

Reply:
After testing for a bit, I can't find an amount to reserve that actually makes a difference… Our "amortized O(1)" complexity for `append(_:)` is pretty real!